improve search in main window to choose case sensitivity

atari2600tim opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
The search box at the bottom of the window now is case insensitive.
#### Motivation for adding to Mudlet
Makes it similar to what people expect
#### Other info (issues closed, discussion etc)
fix #5868

atari2600tim committed:
Update TConsole.cpp

vadi2 reviewed (commented):
This is great but I don't think we want to remove the option of a case-sensitive search at the same time.

Could you add another checkbox to the search menu to be case-sensitive that's off by default?

atari2600tim commented:
At first I thought you meant add a checkbox in this spot in settings.  And I guess add case sensitivity to the code editor, with another checkbox on that tab.  But "to the search menu", I'm thinking you mean something more like VS Code?

MUSHclient and BeipMU and Notepad all have a whole popup dedicated to it so it can be toggled while searching.  Visual Studio Code has an icon `Aa` which goes along with the textbox, and lets you toggle as conveniently as the popups do and might make more sense with the no popup search style.  Having it there all the time like VS Code would take up space. 
 Revealing an icon and resizing or placing it on top of the command line input when you put the cursor in the search box would be a little more complexity and I don't know how it'd do with screen readers, but it would let people toggle it while they're searching, so I'm thinking that might be ideal.  A right click option might also work. 
![image](https://user-images.githubusercontent.com/29287358/193454804-59ccb62a-3f80-49bc-b6a8-e752a6dcdc34.png)

SlySven commented:
How about (for a consistent UI) doing the same as I did for the main search thingy in the editor where a (dummy) action is added to the `QLineEdit` and used to provide a drop down menu to select options (in this case just case-sensitivity) - the icon can then change to indicate the mode that the search is in (in this case either a grey pair of binoculars for case insensitive and a red sort of "Aa" for the opposite)?

vadi2 commented:
What @SlySven said, the way the code editor search does it. What do you think about that?

atari2600tim commented:
I always do Control-F in the code editor and that's why I was suggesting to add caps sensitivity to the code editor too.  Had forgotten there's the other search over on the left, with the binoculars.  That does look good and clear and all.  Will try and do something similar.

Kebap commented:
Ctrl + F searches in the current script. The left-hand search will traverse all scripts, triggers, etc. in the current profile. It has multiple options in dropdown. The main window will search the main output. Should it also receive a button for options? There are many buttons already. However all these searches use different infrastructure, short keys, functionality, etc. Would be nice to align them in the future.

vadi2 commented:
> Should it also receive a button for options?

Not a standalone button but an in-line button, just like in the code editor.

atari2600tim committed:
Merge branch 'Mudlet:development' into timSearchCase

atari2600tim committed:
Merge branch 'Mudlet:development' into timSearchCase

atari2600tim committed:
Merge branch 'Mudlet:development' into timSearchCase

atari2600tim committed:
add menu button into box

atari2600tim committed:
Update TConsole.h

format spaces properly

simplify comments since this won't be doing variables in the future and it's easy to figure out what to add if it ever gets regex or whole word options

vadi2 reviewed (approved):
Looks great to me, thank you!

vadi2 commented:
Is this ready to be merged @atari2600tim?

atari2600tim commented:
I had it as a draft for more discussion but then forgot to renew discussion.

I got it to show the clickable icon in the box and it functions properly, but I did not resize the text area or anything like that.  I don't want to crowd and shink the command input area, but at same time I don't want to like present this with it having like 2 or 3 characters worth of space visible to type in, when it was already small.

Just getting this much figured out took some learning for me though.  I think I will mark it as ready, and if someone would like to make the width dynamic and expand when people are typing in there, or anything like that, they can do a separate PR and I'll just mark this ready.

vadi2 commented:
Sounds good to me. I'll let others review it and merge in a couple of days.

SlySven reviewed (approved):
:+1: Yeah, nicely copied from the main editor search code...!

All I have to do is to fix up the horrible way the search highlighting messes up the original text colouring - I think I can see a way to do that but I just have to find the time...

vadi2 commented:
One idea for improving the UI here would be to add +50 pixels (or the right equivalent for hidpi) when the search box is selected, and decrease them again on focus loss. This'll help keep it normally small and grow bigger when needed.

atari2600tim commented:
It looks like the search box is set to have a range of sizes and then has sizePolicy5 that is earlier defined to mean fixed size, to my vague understanding.
https://github.com/Mudlet/Mudlet/blob/development/src/TConsole.cpp#L416-L418
```cpp
    mpBufferSearchBox->setMinimumSize(QSize(100, 30));
    mpBufferSearchBox->setMaximumSize(QSize(150, 30));
    mpBufferSearchBox->setSizePolicy(sizePolicy5);
```
 and 
```cpp
QSizePolicy sizePolicy5(QSizePolicy::Fixed, QSizePolicy::Fixed);
```

I made a separate branch that resizes when going in and out.  I think I got the event part right, and I set it to one of two widths, but I'm thinking there's some better way that would utilize sizeHint to suggest a size within the max and min.  Or maybe switching between the two is fine but that other area could be rephrased to not have max and min there implying that it is a range.
https://github.com/atari2600tim/Mudlet/commit/1a4c68689470567c30421aba2695725356b5b9a7

vadi2 commented:
Nice 👍 I'll merge this PR since it's good to go, the resizing stuff can be a follow-up PR.
